### PR TITLE
8318422: Allow poller threads be virtual threads

### PR DESCRIPTION
**Pull request by AlanBateman**
Virtual threads doing blocking I/O rely on poller threads to unpark virtual threads when sockets are ready for I/O. Right now, the poller threads are dedicated platform threads that block waiting for I/O events. In some environments it would be better to have the poller thread be virtual threads so that handling of I/O events integrates better with the virtual thread scheduler and doesn't steal cycles from the carrier threads.

The changes to support this are straight forward but it has required refactoring the internal Poller class so the changes might look more than they actually are. The changes mean the Poller implementation supports two modes. The default is changed on Linux to use the "virtual thread poller" mode. The default on other platforms is to use platform threads as before. As part of the refactor, the so-called "indirect" mode with updater threads is removed. This mode came from experimenting in the loom repo a long time ago and probably should have been removed before integration.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318422](https://bugs.openjdk.org/browse/JDK-8318422): Allow poller threads be virtual threads (**Enhancement** - P3)

### Reviewers
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16243/head:pull/16243` \
`$ git checkout pull/16243`

Update a local copy of the PR: \
`$ git checkout pull/16243` \
`$ git pull https://git.openjdk.org/jdk.git pull/16243/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16243`

View PR using the GUI difftool: \
`$ git pr show -t 16243`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16243.diff">https://git.openjdk.org/jdk/pull/16243.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16243#issuecomment-1790905626)

### CODE REVIEW

File: src/java.base/share/classes/sun/nio/ch/Poller.java [hunk 1]
```diff
@@ -24,52 +24,100 @@
...
-    private static final JavaLangAccess JLA = SharedSecrets.getJavaLangAccess();
-    private static final Poller[] READ_POLLERS;
-    private static final Poller[] WRITE_POLLERS;
-    private static final int READ_MASK, WRITE_MASK;
-    private static final boolean USE_DIRECT_REGISTER;
-
-    // true if this is a poller for reading, false for writing
-    private final boolean read;
+abstract class Poller {
+    private static Pollers POLLERS;
```
Comment (Michael-Mc-Mahon) [inline src/java.base/share/classes/sun/nio/ch/Poller.java:44]:
```suggestion
    private static final Pollers POLLERS;
```

File: src/java.base/share/classes/sun/nio/ch/Poller.java [hunk 2]
```diff
@@ -24,52 +24,100 @@
...
-    protected Poller(boolean read) {
-        this.read = read;
+    protected Poller() {
     }
 
     /**
-     * Returns true if this poller is for read (POLLIN) events.
+     * Returns the poller's file descriptor, used when the read and write poller threads
+     * are virtual threads.
+     *
```
Comment (Michael-Mc-Mahon) [inline src/java.base/share/classes/sun/nio/ch/Poller.java:89]: Would it be useful to add some more explanation here like: "When there is no activity on any of a subpoller's descriptors, the descriptor of the subpoller is registered with the master poller" ?

Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/ch/Poller.java:89]: > Would it be useful to add some more explanation here like: "When there is no activity on any of a subpoller's descriptors, the descriptor of the subpoller is registered with the master poller" ?

The moment on Mode.VTHREAD_POLLERS explains this, also links to the Karsten/Barghi where a similar scheme is used. Do you think that's enough?

**Review by Michael-Mc-Mahon — APPROVED** [review]
Looks good to me.

**Comment by AlanBateman** [comment]
> Looks good to me.

Thanks. I've had to one additional update to ProblemList-Virtual.txt to excludevmTestbase/nsk/jdi/VMOutOfMemoryException/VMOutOfMemoryException001/VMOutOfMemoryException001.java on all platforms. It had been excluded on Windows only but it's quite fragile and we tickle it to fail on Linux with these changes. I had a brief exchange with @plummercj in JDK-8285417 about this a few days ago. It's excluded on all platforms in the loom repo already so I should have remembered to include it in this update.

**Review by Michael-Mc-Mahon — APPROVED** [review]
Latest change is good.

**Pull request closed without merging**